### PR TITLE
IR: Allow vector type in atomic load and store

### PR DESCRIPTION
**Pull request by jofrn**
Vector types on atomics are assumed to be invalid by the verifier. However, this type can be valid if it is lowered by codegen.

### CODE REVIEW

**Comment by jofrn** [comment]
#117189 requires this change.

File: llvm/test/Verifier/atomics.ll [hunk 1]
```diff
@@ -1,14 +1,15 @@
 ; RUN: not opt -passes=verify < %s 2>&1 | FileCheck %s
+; CHECK: atomic store operand must have integer, pointer, floating point, or vector type!
+; CHECK: atomic load operand must have integer, pointer, floating point, or vector type!
 
-; CHECK: atomic store operand must have integer, pointer, or floating point type!
-; CHECK: atomic load operand must have integer, pointer, or floating point type!
+%ty = type { i32 };
```
Comment (arsenm) [inline llvm/test/Verifier/atomics.ll:5]: Add to the atomic test in test/Assembler with newly permitted cases

**Comment by jyknight** [comment]
I'm a little confused by the steps being taken here.

Can we have this fallback via cast to integer by default on _all_ platforms, at the same time as making it legal? (That's how floats were handled, e.g.). We shouldn't have this work only on some platforms, but not others.

**Comment by jofrn** [comment]
> I'm a little confused by the steps being taken here.
> 
> Can we have this fallback via cast to integer by default on _all_ platforms, at the same time as making it legal? (That's how floats were handled, e.g.). We shouldn't have this work only on some platforms, but not others.

We can do that. However, the Verifier would still mark it as invalid. The subsequent PR only lowers for X86 ; this PR allows vector types without disabling verifier checks.

**Comment by jyknight** [comment]
I don't understand why do this in multiple steps? Allow vectors in the verifier and update all of the existing implementations of shouldCastAtomic{Load,Story}InIR to cast vectors at the same time?

**Comment by jofrn** [comment]
Verifier checks can be disabled with `-disable-verify`, which already permits the lowering of atomic load/store with vector type in multiple targets, as such, some are already implemented. Permitting the type in Verifier/IR is different than a single implementation of the lowering.

**Comment by arsenm** [comment]
> Can we have this fallback via cast to integer by default on _all_ platforms, at the same time as making it legal? (That's how floats were handled, e.g.). 

Disagree, split the IR verifier change from the codegen pieces.

> We shouldn't have this work only on some platforms, but not others.

There's a ton of IR that will not work on all platforms with varying failure modes, this case isn't special (e.g. 128-bit atomics)

**Comment by nikic** [comment]
I agree with @jyknight that this needs to come with support for all targets, not just X86. Preferably implemented in a way that does not require defining a hook in every single backend.

> > We shouldn't have this work only on some platforms, but not others.
> 
> There's a ton of IR that will not work on all platforms with varying failure modes, this case isn't special (e.g. 128-bit atomics)

We should always aspire to make generic IR constructs work target-independently. If we have fallen short of that in some area, that's not an excuse to repeat the mistake.

**Comment by jyknight** [comment]
> Verifier checks can be disabled with -disable-verify, which already permits the lowering of atomic load/store with vector type in multiple targets

Currently, IR having load/store atomic with a vector type is invalid. If you disable the verifier and provide invalid IR to LLVM, what happens next may be arbitrary -- nobody should be depending on that.

So, I would strongly suggest that the first change be to _unconditionally_ cast the vector to integers -- both in the base class TargetLoweringBase::shouldCastAtomicLoadInIR/shouldCastAtomicStoreInIR and in all the overrides (looks like 3 of those: NVPTX, AMDGPU, SystemZ).

Then, you can make separate follow-up changes for each target to opt out of the cast when possible.

**Comment by arsenm** [comment]
> I agree with @jyknight that this needs to come with support for all targets, not just X86. Preferably implemented in a way that does not require defining a hook in every single backend.

Yes, the codegen patch should implement the generic path for all targets. But in terms of patch splitting, that shouldn't be in this PR

> We should always aspire to make generic IR constructs work target-independently. If we have fallen short of that in some area, that's not an excuse to repeat the mistake.

There are limits to what is practically implementable for operations that support arbitrary types. ppc_f128, x86_fp80, scalable vectors, and arbitrarily sized atomics are never going to be universally supported.

> So, I would strongly suggest that the first change be to unconditionally cast the vector to integers -- both in the base class TargetLoweringBase::shouldCastAtomicLoadInIR/shouldCastAtomicStoreInIR and in all the overrides (looks like 3 of those: NVPTX, AMDGPU, SystemZ).

We really should get rid of these hooks. There's no reason this can't just be handled by ordinary type legalization

**Comment by jyknight** [comment]
> The change from my end only needs X86 support, and that is all that is implemented at the moment

That is insufficient. The functionality (of passing a vector type to load/store atomic) needs to work for all targets, even if your use-case doesn't require it.

**Comment by arsenm** [comment]
> Sounds great, but _that_ certainly seems like a separable change. Since these hooks are how it's done today for floats, I don't see a reason to (initially) do something different for vectors.

I switched AMDGPU to use the DAG legalization in 82bb2534d4de16abb7a51fc646d5c31d6cec5eeb. I thought I switched X86 over but that seems to not have landed 

> The change from my end only needs X86 support. 

There's no plus to only doing x86. Just handle in the default implementation of the hook

**Comment by jofrn** [comment]
Regardless of using the hook, backends are still able to implement atomic load/store of vector, so I do not see why we cannot enable the verifier change separately.

It is very easy to switch all the backends to use the hook, but the DAG legalization change will require more work to check all the backends.

We can add the hook for NVPTX, AMDGPU, and SystemZ by moving the check to the base class. And the DAG legalization can come later since that requires more work.

**Review by nikic — CHANGES_REQUESTED** [review]
We can land the verifier change separately, but only after a non-x86-specific patch for backend support is available as well.

**Comment by jofrn** [comment]
| We can land the verifier change separately, but only after a non-x86-specific patch for backend support is available as well.
Okay, split https://github.com/llvm/llvm-project/pull/111414 into itself and https://github.com/llvm/llvm-project/pull/118793.

**Comment by jofrn** [comment]
Follow-up is here: #120384
@nikic, do you have any other comments?

**Comment by jyknight** [comment]
I'm a little confused as to which PRs are intended to be reviewed now. Is it only the "follow-up" series of 4 PRs you've linked to? Or are some/all of the other 3 PRs you've sent still active?

**Comment by jofrn** [comment]
> I'm a little confused as to which PRs are intended to be reviewed now. Is it only the "follow-up" series of 4 PRs you've linked to? Or are some/all of the other 3 PRs you've sent still active?
The follow-up has a set of new PRs associated. #117625, #111414, #118793, and #120316 are now inactive.

**Pull request closed without merging**